File: back/node/app/core/common/result/Result.ts
type Result<T> = SuccessRes<T> | ErrorRes | NotFound;

type SuccessRes<T> = {
    readonly status: "success";
    readonly data: T;
}

type ErrorRes = {
    readonly status: "error";
    readonly exception: any;
}

type NotFound = {
    readonly status: "notFound";
}

const success = <T>(data: T): SuccessRes<T> => ({ status: "success", data });
const error = (exception: any): ErrorRes => ({ status: "error", exception });
const notFound: NotFound = { status: "notFound" };

export {Result, success, error, notFound}

File: back/kotlin/app/core/common/result/Result.kt
package core.common.result

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onEmpty

sealed interface Result<out T> {
    data class Success<T>(val data: T) : Result<T>
    data class Error(val exception: Throwable) : Result<Nothing>
    object NotFound : Result<Nothing>
}

fun <T> Flow<T>.asResult(): Flow<Result<T>> = map<T, Result<T>> { Result.Success(it) }
    .catch { emit(Result.Error(it)) }
    .onEmpty { emit(Result.NotFound) }